File: frontend/src/api/sectors.ts
import client from './client'
import type { SectorListResponse, Sector, SectorSnapshot } from '@/types'

/**
 * 板块列表。
 *
 * **只读 rank/phase 那几个标签字段时传 false。** 实测完整载荷 2.1MB / 2.75s
 * （304 个板块，每个内嵌全部成员股），是全站最大的一个请求；而它挂在常驻顶栏的
 * useSectorTags 上，**每进任何一个页面都要付一遍**。
 *
 * 带不带成员股的结果形状不同，**必须用不同的 queryKey**——共用一个 key 会让
 * 两类调用方抢同一份缓存，谁先到谁说了算。用 SECTORS_LITE_KEY。
 */
export const fetchSectors = (includeStocks = true) =>
  client.get<SectorListResponse>('/sectors',
    { params: includeStocks ? {} : { include_stocks: false } }).then((r) => r.data)

/** 不带成员股那一份的共享 key。所有只要标签字段的调用方都用它，共享一次请求 */
export const SECTORS_LITE_KEY = ['sectors', 'no-stocks'] as const

export const fetchSector = (code: string) =>
  client.get<Sector>(`/sectors/${code}`).then((r) => r.data)

export const fetchSectorSnapshots = (code: string, days = 30) =>
  client.get<SectorSnapshot[]>(`/sectors/${code}/snapshots`, { params: { days } }).then((r) => r.data)

export interface TopStockItem {
  code: string
  name: string
  pct_today: number | null
  pct_5d: number | null
  pct_10d: number | null
  pct_20d: number | null
  pct_60d: number | null
}

export interface TopStocksResponse {
  bk_code: string
  period: string
  stocks: TopStockItem[]
}

export const fetchSectorTopStocks = (code: string, period = '20d', limit = 10) =>
  client.get<TopStocksResponse>(`/sectors/${code}/top-stocks`, {
    params: { period, limit },
  }).then((r) => r.data)

File: backend/app/routers/sectors.py
from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from typing import List, Optional
from sqlalchemy.orm import Session
from ..database import get_db
from ..models.sector import Sector, SectorDailySnapshot
from ..schemas.sector import SectorListResponse, SectorResponse, SectorDailySnapshotResponse
from ..services.sector_phase_service import get_all_sectors, get_sector_by_code
from ..services.sector_top_stocks_service import get_top_stocks


class TopStockItem(BaseModel):
    code: str
    name: str
    pct_today: Optional[float] = None
    pct_5d: Optional[float] = None
    pct_10d: Optional[float] = None
    pct_20d: Optional[float] = None
    pct_60d: Optional[float] = None


class TopStocksResponse(BaseModel):
    bk_code: str
    period: str
    stocks: List[TopStockItem]

router = APIRouter(prefix="/sectors", tags=["sectors"])


@router.get("", response_model=SectorListResponse)
def list_sectors(db: Session = Depends(get_db)):
    return get_all_sectors(db)


@router.get("/{code}", response_model=SectorResponse)
def get_sector(code: str, db: Session = Depends(get_db)):
    sector = get_sector_by_code(db, code)
    if not sector:
        raise HTTPException(status_code=404, detail=f"Sector {code} not found")
    return sector


@router.get("/{code}/top-stocks", response_model=TopStocksResponse)
def get_sector_top_stocks(
    code: str,
    period: str = Query("20d", regex="^(5d|10d|20d|60d)$"),
    limit: int = Query(10, ge=1, le=50),
):
    """
    返回板块内主板非ST个股，按指定周期涨幅降序排列。
    数据来源：东方财富 clist API（沪市主板 m:1+t:2 + 深市主板 m:0+t:6）。
    """
    stocks = get_top_stocks(bk_code=code, period=period, limit=limit)
    return TopStocksResponse(
        bk_code=code,
        period=period,
        stocks=[TopStockItem(**s.__dict__) for s in stocks],
    )


@router.get("/{code}/snapshots", response_model=list[SectorDailySnapshotResponse])
def get_sector_snapshots(
    code: str,
    days: int = 30,
    db: Session = Depends(get_db),
):
    sector = db.query(Sector).filter(Sector.code == code).first()
    if not sector:
        raise HTTPException(status_code=404, detail=f"Sector {code} not found")

    snaps = (
        db.query(SectorDailySnapshot)
        .filter(SectorDailySnapshot.sector_id == sector.id)
        .order_by(SectorDailySnapshot.date.desc())
        .limit(days)
        .all()
    )
    return list(reversed(snaps))
